File: module_http/src/main/java/cn/roy/zlib/http/retrofit/RetrofitFactory.java
package cn.roy.zlib.http.retrofit;

import android.text.TextUtils;

import java.util.concurrent.TimeUnit;

import cn.roy.zlib.http.core.HttpRequestBaseParam;
import cn.roy.zlib.http.exception.RequestException;
import io.reactivex.annotations.NonNull;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * @Description: Retrofit工厂类
 * @Author: Roy Z
 * @Date: 2019-08-02 16:35
 * @Version: v1.0
 */
public class RetrofitFactory {
    /**
     * 创建Retrofit
     *
     * @param baseUrl
     * @param baseParam
     * @return
     */
    public static Retrofit create(@NonNull String baseUrl, HttpRequestBaseParam baseParam)
            throws RequestException {
        if (baseParam == null || baseParam.getConnectTimeout() < 0
                || baseParam.getWriteTimeout() < 0) {
            throw new RequestException(-1, "超时时间设置有误");
        }
        HttpLoggingInterceptor httpLoggingInterceptor = new HttpLoggingInterceptor(
                new HttpLoggingInterceptor.Logger() {
                    @Override
                    public void log(String message) {
                        System.out.println(message);
                    }
                });
        httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .connectTimeout(baseParam.getConnectTimeout(), TimeUnit.MICROSECONDS)
                .writeTimeout(baseParam.getWriteTimeout(), TimeUnit.MILLISECONDS)
                .readTimeout(baseParam.getReadTimeout(), TimeUnit.MILLISECONDS)
                .callTimeout(baseParam.getCallTimeout() <= 0 ? 0 : baseParam.getCallTimeout(),
                        TimeUnit.MILLISECONDS)
                .addNetworkInterceptor(httpLoggingInterceptor)
                .build();
        Retrofit.Builder retrofitBuilder = new Retrofit.Builder();
        if (!TextUtils.isEmpty(baseUrl)) {
            retrofitBuilder.baseUrl(baseUrl);
        }
        Retrofit retrofit = retrofitBuilder
//                .addConverterFactory(new Retrofit2ConverterFactory())// 装换成JSON
                .addConverterFactory(GsonConverterFactory.create())// 装换成JSON
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .client(okHttpClient)
                .build();

        return retrofit;
    }

}

File: module_http_kotlin/src/main/java/cn/roy/zlib/http/RetrofitFactory.kt
package cn.roy.zlib.http

import android.util.Log
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.CallAdapter
import retrofit2.Converter
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

/**
 * @Description: Retrofit工厂
 * @Author: Roy Z
 * @Date: 2021/02/04
 * @Version: v1.0
 */
object RetrofitFactory {

    @JvmStatic
    fun create(baseUrl: String): Retrofit {
        val configurator = object : RequestConfigurator {
            override fun baseUrl(): String {
                return baseUrl
            }

            override fun connectTimeout(): TimeoutParam {
                return TimeoutParam(5, TimeUnit.SECONDS)
            }

            override fun writeTimeout(): TimeoutParam {
                return TimeoutParam(5, TimeUnit.SECONDS)
            }

            override fun readTimeout(): TimeoutParam {
                return TimeoutParam(5, TimeUnit.SECONDS)
            }

            override fun callTimeout(): TimeoutParam {
                return TimeoutParam(15, TimeUnit.SECONDS)
            }

            override fun logger(): RequestLogger {
                return object : RequestLogger {
                    override fun print(log: String) {
                        Log.d("http", log)
                    }
                }
            }

            override fun requestInterceptors(): List<Interceptor>? {
                return null
            }

            override fun callAdapterFactories(): List<CallAdapter.Factory>? {
                return null
            }

            override fun converterFactories(): List<Converter.Factory>? {
                return null
            }
        }
        return create(configurator)
    }

    @JvmStatic
    fun create(configurator: RequestConfigurator): Retrofit {
        val okHttpClientBuilder = OkHttpClient.Builder()
                .connectTimeout(configurator.writeTimeout().timeout,
                        configurator.writeTimeout().timeUnit)
                .writeTimeout(configurator.writeTimeout().timeout,
                        configurator.writeTimeout().timeUnit)
                .readTimeout(configurator.readTimeout().timeout,
                        configurator.readTimeout().timeUnit)
                .callTimeout(if (configurator.callTimeout().timeout <= 0) 0
                else configurator.callTimeout().timeout,
                        configurator.callTimeout().timeUnit)
        val baseInterceptor = Interceptor { chain ->
            Log.d("roy", "intercept 执行")
            var timestamp = System.currentTimeMillis()
            val httpUrl = chain.request().url
            var uri = httpUrl.toUri()
            val response = chain.proceed(chain.request())
            var cost = System.currentTimeMillis() - timestamp
            configurator.logger().print("<-- 访问：${uri}，请求总耗时：${cost} (ms)-->")
            response
        }
        okHttpClientBuilder.addInterceptor(baseInterceptor)
        if (configurator.requestInterceptors() != null) {
            for (interceptor in configurator.requestInterceptors()!!) {
                okHttpClientBuilder.addInterceptor(interceptor)
            }
        }
        val httpLoggingInterceptor = HttpLoggingInterceptor { message ->
            configurator.logger().print(message)
        }
        httpLoggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
        val okHttpClient = okHttpClientBuilder
                .addNetworkInterceptor(httpLoggingInterceptor)
                .build()
        val retrofitBuilder = Retrofit.Builder().baseUrl(configurator.baseUrl())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .client(okHttpClient)
        if (configurator.callAdapterFactories() != null) {
            for (callAdapterFactory in configurator.callAdapterFactories()!!) {
                retrofitBuilder.addCallAdapterFactory(callAdapterFactory)
            }
        }
        if (configurator.converterFactories() != null) {
            for (converterFactory in configurator.converterFactories()!!) {
                retrofitBuilder.addConverterFactory(converterFactory)
            }
        } else {
            retrofitBuilder.addConverterFactory(GsonConverterFactory.create())
        }
        return retrofitBuilder.build()
    }

}